### PR TITLE
Text splitting initial implementation

### PR DESCRIPTION
**Pull request by SpirosMakris**
This is an initial implementation of text splitting, including token overlap.
I'm not sure if this is, at all, similar to what you had in mind.

NaiveWhitespaceTextSplitter  probably isn't very useful. Can either be removed or augmented with sentence end punctuation splitting.

Also,I didn't include any tests for LLama versions, because it seems tokenizing needs the actual model and the ci/cd would probably fail at those tests?

This is my first PR ever, so if something looks wrong, it probably is wrong.
Corrections/remarks are welcome

### CODE REVIEW

File: llm-chain-openai/src/chatgpt/text_splitter.rs [hunk 1]
```diff
@@ -0,0 +1,122 @@
+use super::step::Model;
+use llm_chain::text_splitter::{TextSplitError, TextSplitter};
+use tiktoken_rs::CoreBPE;
+
+pub struct OpenAITextSplitter {
+    model: Model,
+}
+
+impl OpenAITextSplitter {
```
Comment (williamhogman) [inline llm-chain-openai/src/chatgpt/text_splitter.rs:9]: If you could make: 

https://github.com/sobelio/llm-chain/blob/09f698af41be963d1c3d4f481a1ef656c5f1ac9d/llm-chain-openai/src/chatgpt/executor.rs#L71-L82

Use this system that would be great :)

File: llm-chain/src/text_splitter.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+#[derive(Error, Debug)]
+pub enum TextSplitError {
+    #[error("Error tokenizing input text")]
+    TokenizationError,
+    #[error("Error converting token chunk to string")]
+    ChunkToStringError,
+}
+
+pub trait TextSplitter<TokenType: Clone> {
```
Comment (williamhogman) [inline llm-chain/src/text_splitter.rs:12]: Same things with 
https://github.com/sobelio/llm-chain/blob/09f698af41be963d1c3d4f481a1ef656c5f1ac9d/llm-chain/src/tokens.rs#L37-L83

Can we make this use you system

**Review by williamhogman — COMMENTED** [review]
https://github.com/sobelio/llm-chain/blob/09f698af41be963d1c3d4f481a1ef656c5f1ac9d/llm-chain/src/traits.rs#L101-L132

If we could make the existing set up work with your system that would be amazing :)

**Comment by williamhogman** [comment]
Really good PR let's make our existing systems use this 🎉

**Comment by SpirosMakris** [comment]
Hey Will! Thanks!

I've split the tokenizer to tokens/to string into it's own trait, rebases on current main and now working on integrating it with the Executors

**Comment by williamhogman** [comment]
@SpirosMakris seems like the right direction :) Keep up the great work :)

**Review by williamhogman — APPROVED** [review]
Nice really like the new version

**Comment by SpirosMakris** [comment]
Hey Will, thanks!

Should i continue with replacing the splitting functionality in `split_at_tokens()` & `split_to_fit()` in `ExecutorTokenCountExt`?

**Comment by williamhogman** [comment]
> Hey Will, thanks!
> 
> Should i continue with replacing the splitting functionality in `split_at_tokens()` & `split_to_fit()` in `ExecutorTokenCountExt`?

Yes please!

**Pull request merged**